### PR TITLE
timesyncd: split out port number when given domain:port, I.P.v.4:port or [IP:v:6]:port configuration.

### PR DESCRIPTION
**Pull request by k3ninho**
timesyncd attempts to resolve an IPv4 address like 10.0.0.1:1234 or IPv6 address like [fe80::1]:1234 via DNS lookup, when the library functions have already added port information to the sockaddr union info within the timesyncd-manager's address data. If the name of the server is IPv4 and contains ':' or the name has ']' followed by ':' (as in the above examples), then use the supplied value from the sockaddr.

The crucial log line was:
`Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Resolving 192.168.1.253:1234...`

I found this because a local override on a Debian system failed to connect to an old cellphone running [PublicNTP's Time Server App](https://github.com/PublicNTP/time-server-app) that was serving GPS time data on on 192.168.1.253:1234.  The configuration and state are shown below, along with timedatectl's inability to read a server name from this supplied value and debug logs showing that it's attempting to resolve DNS for 192.168.1.253:1234. (It's a question for another day why there's no log content indicating that the fallback servers are used.)
```

user@test_system:~$ sudo timedatectl timesync-status
   Server: 77.104.162.218 (0.debian.pool.ntp.org)
Poll interval: 17min 4s (min: 32s; max 34min 8s)
 Leap: normal
  Version: 4
  Stratum: 1
Reference: PPS
Precision: 1us (-24)
Root distance: 22us (max: 5s)
   Offset: -190us
Delay: 23.129ms
   Jitter: 1.455ms
 Packet count: 6
Frequency: +2.364ppm
user@test_system:~$ sudo nano /etc/systemd/timesyncd.conf.d/80-local.conf 
user@test_system:~$ cat /etc/systemd/timesyncd.conf.d/80-local.conf 
# local gps item
[Time]
NTP=192.168.1.253:1234
user@test_system:~$ sudo systemctl daemon-reload
user@test_system:~$ systemctl restart systemd-timesyncd 
user@test_system:~$ sudo timedatectl timesync-status
   Server: n/a (192.168.1.253:1234)
Poll interval: 0 (min: 32s; max 34min 8s)
 Packet count: 0
```

Full debug log with 192.168.1.253:1234 configuration: 
```
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Added new system server 192.168.1.253:1234. 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Added new fallback server 0.debian.pool.ntp.org. 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Added new fallback server 1.debian.pool.ntp.org. 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Added new fallback server 2.debian.pool.ntp.org. 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Added new fallback server 3.debian.pool.ntp.org. 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: systemd-timesyncd running as pid 31868 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Selected server 192.168.1.253:1234. 
Apr 08 12:14:13 test_system systemd[1]: Started systemd-timesyncd.service - Network Time Synchronization. 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Resolving 192.168.1.253:1234... 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Bus bus-api-timesync: changing state AUTHENTICATING → HELLO Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RequestName cookie=2 reply_cookie=0 signature=su error-name=n/a error-message=n/a 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.600 path=n/a interface=n/a member=n/a  cookie=1 reply_cookie=1 signature=s error-name=n/a error-message=n/a 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Bus bus-api-timesync: changing state HELLO → RUNNING 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Got message type=signal sender=org.freedesktop.DBus.Local destination=n/a path=/org/freedesktop/DBus/Local interface=org.freedesktop.DBus.Local member=Connected  cookie=4294967295 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Got message type=signal sender=org.freedesktop.DBus destination=:1.600 path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameAcquired  cookie=2 reply_cookie=0 signature=s error-name=n/a error-message=n/a 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Got message type=signal sender=org.freedesktop.DBus destination=:1.600 path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameAcquired  cookie=3 reply_cookie=0 signature=s error-name=n/a error-message=n/a 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.600 path=n/a interface=n/a member=n/a  cookie=4 reply_cookie=2 signature=u error-name=n/a error-message=n/a 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: varlink: Setting state idle-client 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: varlink: Sending message: {"method":"io.systemd.Resolve.ResolveHostname","parameters":{"name":"192.168.1.253:1234","flags":1024}} 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: varlink: Changing state idle-client → calling 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: Successfully acquired requested service name. 
Apr 08 12:14:13 test_system systemd-timesyncd[31868]: varlink: New incoming message: {"error":"io.systemd.Resolve.DNSError","parameters":{"rcode":3}}
```

The test-timesyncd items only check that the strings parse OK as domain names, but the underlying functions accept IP addresses so I've added 10.0.0.1:1234 and [fe80::1]:1234 and variants like them.

```
================================== 799/1473 ==================================
test:         systemd:timesync / test-timesync
start time:   18:26:48
duration:     0.02s
result:       exit status 0
command:      SYSTEMD_LANGUAGE_FALLBACK_MAP=/home/k3ninho/linux/systemd/src/locale/language-fallback-map PROJECT_BUILD_ROOT=/home/k3ninho/linux/systemd/build LD_LIBRARY_PATH=/home/k3ninho/linux/systemd/build/src/shared MALLOC_PERTURB_=215 PATH=/home/k3ninho/linux/systemd/build:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games SYSTEMD_SLOW_TESTS=0 /home/k3ninho/linux/systemd/build/test-timesync
----------------------------------- stderr -----------------------------------
/* test_manager_parse_string */
Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
Failed allocate memory pressure event source, ignoring: Permission denied
Failed to get link NTP servers: No such file or directory
Added new fallback server time1.google.com.
Added new fallback server time2.google.com.
Added new fallback server time3.google.com.
Added new fallback server time4.google.com.
Added new system server time1.foobar.com.
Added new system server time2.foobar.com.
Invalid NTP server name or address, ignoring: axrfav.,avf..ra
Added new system server time2.foobar.com:1234.
Added new system server 10.0.0.1.
Added new system server fe80::1.
Added new system server [fe80::1].
Added new system server 10.0.0.1:1234.
Added new system server [fe80::1]:1234.
Invalid NTP server name or address, ignoring: 12345..123
Added new fallback server time1.foobar.com.
Added new fallback server time2.foobar.com.
Invalid NTP server name or address, ignoring: axrfav.,avf..ra
Added new fallback server time2.foobar.com:1234.
Added new fallback server 10.0.0.1.
Added new fallback server fe80::1.
Added new fallback server [fe80::1].
Added new fallback server 10.0.0.1:1234.
Added new fallback server [fe80::1]:1234.
Invalid NTP server name or address, ignoring: 12345..123
Invalid NTP server name or address, ignoring: axrfav.,avf..ra
Added new link server fe80::1].
Invalid NTP server name or address, ignoring: 12345..123
Removed server time1.foobar.com.
Removed server time2.foobar.com.
Removed server time2.foobar.com:1234.
Removed server 10.0.0.1.
Removed server fe80::1.
Removed server [fe80::1].
Removed server 10.0.0.1:1234.
Removed server [fe80::1]:1234.
Removed server fe80::1].
Removed server time1.google.com.
Removed server time2.google.com.
Removed server time3.google.com.
Removed server time4.google.com.
Removed server time1.foobar.com.
Removed server time2.foobar.com.
Removed server time2.foobar.com:1234.
Removed server 10.0.0.1.
Removed server fe80::1.
Removed server [fe80::1].
Removed server 10.0.0.1:1234.
Removed server [fe80::1]:1234.
==============================================================================

================================== 924/1473 ==================================
test:         systemd:test / test-dns-domain
start time:   18:28:48
duration:     0.01s
result:       exit status 0
command:      SYSTEMD_LANGUAGE_FALLBACK_MAP=/home/k3ninho/linux/systemd/src/locale/language-fallback-map PROJECT_BUILD_ROOT=/home/k3ninho/linux/systemd/build LD_LIBRARY_PATH=/home/k3ninho/linux/systemd/build/src/shared MALLOC_PERTURB_=152 PATH=/home/k3ninho/linux/systemd/build:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games SYSTEMD_SLOW_TESTS=0 /home/k3ninho/linux/systemd/build/test-dns-domain
----------------------------------- stderr -----------------------------------
/* test_dns_label_unescape */
...
/* test_dns_name_is_valid_or_address */
/* test_dns_name_dot_suffixed */
==============================================================================
```

---
On branch timesyncd-parse-and-use-addresses-with-port-number 
Your branch is up to date with 'origin/main'.

Changes to be committed:
	modified:   src/test/test-dns-domain.c
	modified:   src/timesync/test-timesync.c
	modified:   src/timesync/timesyncd-manager.c
	modified:   src/timesync/timesyncd-server.h

### CODE REVIEW

**Comment by k3ninho** [comment]
TL;DR
My changes to src/timesync/timesync-server.c show memory leaks in the ASAN/UBSAN tests:
- 8x in `greedy_realloc()` at `src/basic/alloc-util.c:70` used by `extract_first_word()` at `src/basic/extract-word.c:62`
- 3x in `strdup()` use by `server_name_new()` while initialising ServerName struct at `src/timesync/timesyncd-server.c:97`

File: src/test/test-dns-domain.c [hunk 1]
```diff
@@ -734,8 +734,10 @@ TEST(dns_name_is_valid_or_address) {
         assert_se(dns_name_is_valid_or_address("foobar..com") == 0);
         assert_se(dns_name_is_valid_or_address("foobar.com.") > 0);
         assert_se(dns_name_is_valid_or_address("127.0.0.1") > 0);
+        assert_se(dns_name_is_valid_or_address("127.0.0.1:8080") > 0);
         assert_se(dns_name_is_valid_or_address("::") > 0);
         assert_se(dns_name_is_valid_or_address("::1") > 0);
+        assert_se(dns_name_is_valid_or_address("[::1]:8080") > 0);
```
Comment (poettering) [inline src/test/test-dns-domain.c:740]: uh, how could that possibly succeed? this is weird. the function is supposed to valid DNSnames or IP addresses, but not a port number

File: src/timesync/test-timesync.c [hunk 2]
```diff
@@ -13,16 +14,65 @@ TEST(manager_parse_string) {
...
+        ASSERT_TRUE(manager_parse_server_string(m, SERVER_LINK, "time1.foobar.com time2.foobar.com axrfav.,avf..ra time2.foobar.com:1234 10.0.0.1 fe80::1 [fe80::1] 10.0.0.1:1234 [fe80::1]:1234 12345..123") == 0);
+}
+
+TEST(server_name_parse_port) {
+        /* When supplied with a port number, use it (and re-use it if we fail to connect), and parse all of
+         * host.domain, host.domain:port, I.P.v.4, IP::v:6, [IP::v:6], I.P.v.4:port, [IP::v:6]:port appropriately. */
+        _cleanup_(server_name_freep) ServerName *n = NULL;
+        n = new(ServerName, 1);
+        *n = (ServerName) {
+                .string = strdupa_safe("time1.foobar.com"),
```
Comment (poettering) [inline src/timesync/test-timesync.c]: uh, this is a memory corruption really. you allocate this on the stack, but once server_name_free() gets invoked on this we'll free it with free().

File: src/timesync/timesyncd-server.h [hunk 3]
```diff
@@ -34,6 +34,7 @@ struct ServerName {
 
         ServerType type;
         char *string;
+        char *overridden_port;
```
Comment (poettering) [inline src/timesync/timesyncd-server.h]: hmm, the port should really be a uint16_t

File: src/timesync/test-timesync.c [hunk 4]
```diff
@@ -13,16 +14,65 @@ TEST(manager_parse_string) {
...
+                .overridden_port = NULL,
+        };
+        ASSERT_NULL(n->overridden_port); //no override
```
Comment (poettering) [inline src/timesync/test-timesync.c]: as per coding style: we do not use // for explanatory comments, but use /* */ for that. 

// comments are left for FIXME style comments

File: src/timesync/test-timesync.c [hunk 5]
```diff
@@ -13,16 +14,65 @@ TEST(manager_parse_string) {
...
+
+        ASSERT_TRUE(server_name_parse_port(n) == 0);
+        ASSERT_STREQ(n->string, "time1.foobar.com");
+        ASSERT_NULL(n->overridden_port); //no override
+
+        n->string = strdupa_safe("8.8.8.8");
```
Comment (poettering) [inline src/timesync/test-timesync.c]: use `assert_se(free_and_strdup(&n->string, "8.8.8.8") >= 0)`.

i.e. use heap for this

File: src/timesync/timesyncd-server.c [hunk 6]
```diff
@@ -175,3 +178,39 @@ void server_name_flush_addresses(ServerName *n) {
...
+        const char *sqr = "]", *col = ":";
+        char *ret_sqr, *ret_col, *last_col;
+
+        ret_sqr = strrchr(n->string, *sqr);
+        ret_col = strchr(n->string, *col);
+        last_col = strrchr(n->string, *col);
+
+        if (ret_sqr == NULL && ret_col == NULL) { // server.domain or I.P.v.4
+                return 0;
+        } else if (ret_sqr == NULL && strlen(ret_col) == strlen(last_col)) { // has no ']' and exactly one ":"
```
Comment (poettering) [inline src/timesync/timesyncd-server.c]: as per coding style we do not use {} on single line if blocks.

as per coding style we do not compare explicitly with == NULL, but rely on C's downgrade-to-bool feature for pointer non-null checks

File: src/timesync/timesyncd-server.c [hunk 7]
```diff
@@ -175,3 +178,39 @@ void server_name_flush_addresses(ServerName *n) {
...
+                word = last_col;
+                (void) extract_first_word(&word, &n->overridden_port, col, 0);
+                log_debug("Matched [IP::v:6]:port, output  %s / %s", n->string, n->overridden_port);
+                return 2;
+        } else if (ret_sqr != NULL && last_col != NULL && strlen(ret_sqr) != strlen(last_col)+1) {
+                // [IP::v:6] without port -- no "]:" substring
+                return 0;
+        } else {
+                assert_not_reached();
+                return -1;
```
Comment (poettering) [inline src/timesync/timesyncd-server.c]: do not make up errors, we return negative Exyz errors, errno style.

Comment (poettering) [reply src/timesync/timesyncd-server.c]: also, assert that some piece of code is not reached followed by code doesn't really make much sense.

File: src/timesync/timesyncd-server.c [hunk 8]
```diff
@@ -175,3 +178,39 @@ void server_name_flush_addresses(ServerName *n) {
...
+
+        ret_sqr = strrchr(n->string, *sqr);
+        ret_col = strchr(n->string, *col);
+        last_col = strrchr(n->string, *col);
+
+        if (ret_sqr == NULL && ret_col == NULL) { // server.domain or I.P.v.4
+                return 0;
+        } else if (ret_sqr == NULL && strlen(ret_col) == strlen(last_col)) { // has no ']' and exactly one ":"
+                const char *word = strdupa_safe(n->string);
+                (void) extract_first_word(&word, &n->string, col, 0);
```
Comment (poettering) [inline src/timesync/timesyncd-server.c]: this leaks memory, as old n->string is not freed.

File: src/timesync/timesyncd-server.c [hunk 9]
```diff
@@ -175,3 +178,39 @@ void server_name_flush_addresses(ServerName *n) {
...
+                (void) extract_first_word(&word, &n->overridden_port, col, 0);
```
Comment (poettering) [inline src/timesync/timesyncd-server.c]: i really don't like that n->string changes contents once server_name_parse_port() is invoked.

we want that our fields are reasonably normalized, they contain the same type of data all the time.

the port number should already be parsed the moment we read it from the config file, and never be maintained in server_name_new() as part of the server string, it should be split off there already.

File: src/timesync/timesyncd-manager.c [hunk 10]
```diff
@@ -876,7 +879,27 @@ int manager_connect(Manager *m) {
...
 
-                r = resolve_getaddrinfo(m->resolve, &m->resolve_query, m->current_server_name->string, "123", &hints, manager_resolve_handler, NULL, m);
+                /* port 123 used to be hard-coded, while config can be fed an ip address overriding the port number,
+                 * eg server.domain:1234, 10.0.0.1:1234 or [fe80::1]:1234
+                 * plus, if the name is [IP::v:6] in brackets, resolve won't accept it unles we remove the brackets
+                 * r = resolve_getaddrinfo(m->resolve, &m->resolve_query, m->current_server_name->string, "123", &hints, manager_resolve_handler, NULL, m); */
+                addr = m->current_server_name->string;
+                if (strchr(addr, *sqo) != NULL && strcspn(addr, sqo) == 0
+                                               && strlen(addr)-1 == strcspn(addr, sqc)) {
+                        temp = strdup(m->current_server_name->string);
```
Comment (poettering) [inline src/timesync/timesyncd-manager.c]: oom check

File: src/timesync/timesyncd-manager.c [hunk 11]
```diff
@@ -876,7 +879,27 @@ int manager_connect(Manager *m) {
...
+                        (void) extract_first_word(&temp, &addr, sqc, 0);
+                        temp = strdup(addr);
```
Comment (poettering) [inline src/timesync/timesyncd-manager.c]: temp is leaked

**Review by poettering — CHANGES_REQUESTED** [review]
please make sure that the patch is free of oom issues. Also, please have a look at our coding style:

https://systemd.io/CODING_STYLE/

File: src/timesync/timesyncd-server.h (hunk 3 above)
Comment (k3ninho) [reply src/timesync/timesyncd-server.h]: Noted. I'm going to take a different approach following feedback below so this is moot, but I chose to use a string because it's user-supplied configuration as a string and goes out to `resolve_getaddrinfo()` as a string.

File: src/timesync/timesyncd-server.c (hunk 9 above)
Comment (k3ninho) [reply src/timesync/timesyncd-server.c]: The lifecycle runs in such a way that `timesyncd-conf.c` has `manager_parse_server_string()` which calls `timesyncd-server.c`'s `server_name_new()` to create the `ServerName` struct. Then the `ServerAddress` struct where you might store the port number is not created until `manager_connect()`  calls resolve_getaddrinfo() with `m->current_server_name->string` as the server address.

Consequently, I'll move this so that `server_name_new()` doesn't do anything (even removing the changes to the `ServerAddress` struct), then using `manager_connect()` in `timesyncd-manager.c` to parse the `m->current_server_name->string` to see if it has port information which we can use appropriately. (It's a better match to the manager using `server_name_flush_addresses()` just before calling `resolve_getaddr_info()`).

File: src/timesync/timesyncd-manager.c (hunk 10 above)
Comment (k3ninho) [reply src/timesync/timesyncd-manager.c]: Noted. I will address this.

File: src/timesync/timesyncd-manager.c (hunk 11 above)
Comment (k3ninho) [reply src/timesync/timesyncd-manager.c]: Noted. I will address this.

File: src/timesync/timesyncd-server.c (hunk 8 above)
Comment (k3ninho) [reply src/timesync/timesyncd-server.c]: Noted. I will address this.

File: src/timesync/timesyncd-server.c (hunk 7 above)
Comment (k3ninho) [reply src/timesync/timesyncd-server.c]: Noted. I will address this.

File: src/timesync/timesyncd-server.c (hunk 6 above)
Comment (k3ninho) [reply src/timesync/timesyncd-server.c]: Noted. I will address this.

File: src/timesync/test-timesync.c (hunk 5 above)
Comment (k3ninho) [reply src/timesync/test-timesync.c]: Noted. I will address this.

File: src/timesync/test-timesync.c (hunk 4 above)
Comment (k3ninho) [reply src/timesync/test-timesync.c]: Noted. I will address this.

File: src/timesync/test-timesync.c (hunk 2 above)
Comment (k3ninho) [reply src/timesync/test-timesync.c]: Noted. I will address this.

File: src/test/test-dns-domain.c (hunk 1 above)
Comment (k3ninho) [reply src/test/test-dns-domain.c:740]: Thank you for your feedback. I hope it makes this MR better and into a state you can accept.

The underlying call to glibc's inet_pton() thinks this is valid. Maybe `manager_parse_server_string` in `src/timesync/timesyncd-conf.c` should reject a string containing one colon (for server.domain:port or I.P.v.6:port) or substring ']:' (for [IP::v:6]:port), but instead you got this MR and attempt to parse and handle non-standard port numbers.

This configuration flowing through to [manager_connect:L880](https://github.com/systemd/systemd/blob/main/src/timesync/timesyncd-manager.c#L880) was why I raised #32218.

**Comment by k3ninho** [comment]
Thanks @bluca @yuwata @poettering, I'm going to withdraw this MR.
